src/shared/exec-util.c

```diff
@@ -567,17 +566,18 @@ int _fork_agent(const char *name, const int except[], size_t n_except, pid_t *re
 
         /* In the child: */
 
-        stdout_is_tty = isatty_safe(STDOUT_FILENO);
-        stderr_is_tty = isatty_safe(STDERR_FILENO);
+        bool stdin_is_tty = isatty_safe(STDIN_FILENO);
+        bool stdout_is_tty = isatty_safe(STDOUT_FILENO);
+        bool stderr_is_tty = isatty_safe(STDERR_FILENO);
 
-        if (!stdout_is_tty || !stderr_is_tty) {
+        if (!stdin_is_tty || !stdout_is_tty || !stderr_is_tty) {
                 int fd;
 
                 /* Detach from stdout/stderr and reopen /dev/tty for them. This is important to ensure that
                  * when systemctl is started via popen() or a similar call that expects to read EOF we
                  * actually do generate EOF and not delay this indefinitely by keeping an unused copy of
                  * stdin around. */
-                fd = open("/dev/tty", O_WRONLY);
+                fd = open("/dev/tty", stdin_is_tty ? O_WRONLY : O_RDWR);
```

Comment:
shouldnt we also use O_RDONLY here if only input is needed? just to be fully correct?